Feature/832 java app automation

lfronczyk opened the pull request.

lfronczyk committed:
AB#832 - walking through java app ui elements (parsing string path, traversing the path with real elements, loading data from java ui elements, performing actions)

lfronczyk committed:
AB#832 - updated tests to match new rule that the path must contain at last two elements with number (jvm id) or wildcard (any jvm id) as a first element

lfronczyk committed:
AB#832 - tests moved to more proper place + added tests for wildcard

lfronczyk committed:
AB#832 - addon structure base

lfronczyk committed:
AB#832 - refactoring at PathElement

lfronczyk committed:
AB#832 - initial implementation of control tree view (panel) & related changes + refactoring

lfronczyk committed:
AB#832 - panel filled with java controls (AccessibleNodes aka nodes) tree
- first attempt to generate path to a node
- JPathService renamed to PathService
- fucked up non-awaitable enumeration of java windows `Windows_run()` (per instance of AccessBridge!) that takes some random time -> wrapped up in a pseudofactory, so there's just one instance that is properly initialized only once

lfronczyk committed:
AB#832 - finally got this shitty JAB working inside Robot (both displays the tree and does that at the Robot start)

lfronczyk committed:
AB#832 - PathParser renamed to Services

lfronczyk committed:
AB#832 - fixed namespaces

lfronczyk committed:
AB#832 - click and doaction commands

lfronczyk committed:
AB#832 - gettext, settext commands + draft of NodeStructure (some basic properties of JavaAccessibleNode described in NodeModel)

lfronczyk committed:
AB#832 - refactor (renamed structures files)

lfronczyk committed:
AB#832 - refactor (renamed structures class names)

lfronczyk committed:
AB#832 - changed list types, added Path property

lfronczyk committed:
AB#832 - added javaui.getrectangle, waitfor.javaui + removed unused UIElement

lfronczyk committed:
dependencies to PathService => IPathService

lfronczyk committed:
AB#832 - JavaUIGetRectangleCommand added to project

lfronczyk committed:
AB#832 - an attempt to add focusing functionality (requestFocus function present in JAB native dlls but missing in the interop)

lfronczyk committed:
AB#832 - RequestFocus used in command (JavaUISetFocusCommand)

lfronczyk committed:
AB#832 - AB factory moved to a bit more proper place

lfronczyk committed:
AB#832 - direct interop method calls moved from model to service (model calls service)

lfronczyk committed:
AB#832 - tooltip with node details for tree of controls + removed Path property

lfronczyk committed:
AB#832 - proper name for folder with models

lfronczyk committed:
AB#832 - proper namespace for NodeModel

lfronczyk committed:
AB#832 - a draft of xpath handling (with refactored version of XPathUIElementBuilder from G1ANT.Addon.UI that implements "position()" function)

lfronczyk committed:
AB#832 - code cleanup

lfronczyk committed:
AB#832 - NodeProperty & XPathUIElementBuilder moved to proper folders

lfronczyk committed:
AB#832 - fixed namespaces

lfronczyk committed:
AB#832 - Path replaced with XPath

lfronczyk committed:
AB#832 - XPath function "position()" is now more standard-compatible - it returns an index and takes no parameters (use as ie. ‴/ui[position()=0]‴)

lfronczyk committed:
AB#832 - XPath "position()" function works now with all int comparison operators (=, >, <, <=, >=)

lfronczyk committed:
removed unused folder inside project

lfronczyk committed:
AB#854 - single-file addon: load JAB libraries from resources
- constructor uses parameter containing source folder for JAB libraries + fixed formatting
- all dependencies embedded into assembly
- extract non-assembly files into "add-on" directory before JAB initialization

lfronczyk committed:
AB#854 - refactor - AccessBridgeFactory contacts with outer world through services

lfronczyk committed:
AB#854 - further refactoring

lfronczyk committed:
AB#854 - fixed bug that caused display of random mess at available action list

lfronczyk committed:
AB#854 - fixed bug that caused Bounds to be empty

lfronczyk committed:
AB#854 - added highlight node functionality (through context menu or button above the nodes tree)

lfronczyk committed:
AB#854 - refactoring
- logic moved from view (Panel) to new controller
- marker form code moved to new class (MarkerForm)

lfronczyk committed:
AB#832 - removed post-build command

lfronczyk committed:
AB#832 - removed obsolete stuff

lfronczyk committed:
AB#832 - removed post-build command

lfronczyk committed:
AB#832 - embed WindowsAccessBridgeInterop from $(Configuration), not Debug

lfronczyk committed:
AB#832 - added javaui.mouseclick to emulate "click" (with the mouse) on the controls that do not handle "click" action

lfronczyk committed:
AB#832 - javaui.gettext now returns text of control (previously it was its name)

lfronczyk committed:
AB#832 - added a parameter to javaui.setfocus that allows to bring the app window to top before applying focus

lfronczyk committed:
AB#832 - too long tooltip lines are broken into more lines + added option to tree context menu to copy the tooltip to clipboard

lfronczyk committed:
AB#832 - DRY - shared the code that brings node window to front + highlight brings window to top

lfronczyk committed:
AB#832 - BringToFront argument of javaui.setfocus made optional

lfronczyk committed:
AB#832 - default value for BringToFront argument of javaui.setfocus

lfronczyk committed:
AB#832 - removed jvm details from the xpath, now it is present as /* + xpath builder handles /*, /., //, /.. and so on

lfronczyk committed:
AB#832 - path to jvm window is generated using descendant::, because /* does not work as expected (returns just a first element from collection) and implementing it requires lots of work

lfronczyk committed:
AB#832 - JavaNodeStructure removed in favor of json + refactored lazy evaluator of NodeModel.Actions

lfronczyk committed:
AB#832 - fixed a typo in .csproj

lfronczyk committed:
AB#832 - resource extraction (files required by AccessBridgeFactory to be stored on the disk before Java Access Bridge is initialized) moved to separate service for the sake of SRP

lfronczyk committed:
AB#832 - unit test for resource extraction + some required refactoring and bug fixing

lfronczyk committed:
AB#832 - unit tests for resource extraction

pwitusik commented on G1ANT.Addon.JavaUI/Structures/JavaNodeStructure.cs:
```diff
@@ -0,0 +1,91 @@
+    {
+        public JavaNodeStructure(object value, string format = null, AbstractScripter scripter = null)
+            : base(value, format, scripter)
+        { }
```
Add all indexes supported by structure to Indexes property

pwitusik commented on G1ANT.Addon.JavaUI/Commands/JavaUIGetNodeCommand.cs:
```diff
@@ -0,0 +1,40 @@
+namespace G1ANT.Addon.UI
+{
+    [Command(Name = "javaui.getnode",
+        Tooltip = "This command performs an action at a desktop Java application UI element specified by JPath structure")]
```
Wrong tooltip

pwitusik commented on G1ANT.Addon.JavaUI/Commands/JavaUIDoActionCommand.cs:
```diff
@@ -0,0 +1,33 @@
+            [Argument(Required = true, Tooltip = "Desktop Java application UI element to be clicked")]
+            public JavaPathStructure Path { get; set; }
+
+            [Argument(Required = true, Tooltip = "Desktop Java application UI element to be clicked")]
```
Wrong tooltip

pwitusik commented on G1ANT.Addon.JavaUI/Commands/JavaUIGetTextCommand.cs:
```diff
@@ -0,0 +1,38 @@
+namespace G1ANT.Addon.UI
+{
+    [Command(Name = "javaui.gettext",
+        Tooltip = "This command performs an action at a desktop Java application UI element specified by JPath structure")]
```
Wrong tooltip

lfronczyk committed:
AB#832 - unit tests, bug fixes and optimizations for NodeModel:
- call to nodeService.GetActions performed only, if info contains data that there are some actions
- fixed xpath by position() ("@position()" => "position()")
- unit tests for all methods and fields of NodeModel (with Accessible CN/Jvm/Window variations)

lfronczyk committed:
AB#832 - formating + elements required for mocking made virtual

lfronczyk committed:
removed code commented out

lfronczyk committed:
AB#832 - removed duplicated Windows_run() call (already called during initialization of AccessBridge interop)

lfronczyk committed:
AB#832 - fixed command namespaces

lfronczyk committed:
AB#832 - fixed command namespaces, again :)

lfronczyk committed:
AB#832 - embedded project taken from Release :/ (can't use $(Configuration) in .resx) in order to fix build @ Azure

lfronczyk committed:
AB#832 - fixed Release configuration of solution (somehow it had Debug projects assigned)

lfronczyk committed:
AB#832 - fixed all command descriptions

lfronczyk committed:
AB#832 - patch for DI process (LoadMetadata at Azure Pipelines caused Panel creation, which caused JABFactory to initialize JAB with not initialized SettingsContainer => NRE)

lfronczyk committed:
AB#832 - typo in description

lfronczyk committed:
AB#832 - missing description
